From c5d0aee1568a48a2490f1b96e039ad9f42866bd9 Mon Sep 17 00:00:00 2001
From: Fei Ma <quic_mafeng@quicinc.com>
Date: Thu, 5 Sep 2024 13:40:20 +0800
Subject: [PATCH 04/11] pinctrl: msm: Don't call irq_chip_eoi_parent if
 interrupts routed via dirconn

A crash issue occurs in this scenario:
LV.3.2.1 metal + coqos hypervisor, using TLMM GPIO that maps to PDC,
irq trigger type is dual edge.

We found a falling edge event which interrupt routed via direct connection
once to request irq. During irq handle, It will call irq_chip_eoi_parent,
which invloked an invalid EOI to GIC.
	gic_handle_irq
	  __handle_domain_irq
	    msm_gpio_dirconn_handler
              generic_handle_irq
                handle_fasteoi_irq
                  irq_chip_eoi_parent

In coqos hvpervisor virtualization, every invalid EOI is documented
as unpredictable behavior in the hardware. The invalid EOI before
the valid one causes the hardware to ignore the subsequent valid one.
That's the root cause of this problem. So consider don't call irq_chip
_eoi_parent if interrupts routed via dirconn to avoid invoking invalid
EOI.

Signed-off-by: Fei Ma <quic_mafeng@quicinc.com>
---
 drivers/pinctrl/qcom/pinctrl-msm.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/drivers/pinctrl/qcom/pinctrl-msm.c b/drivers/pinctrl/qcom/pinctrl-msm.c
index c626c9b4d..d68f396c0 100644
--- a/drivers/pinctrl/qcom/pinctrl-msm.c
+++ b/drivers/pinctrl/qcom/pinctrl-msm.c
@@ -91,6 +91,7 @@ struct msm_pinctrl {
 };
 
 static struct msm_pinctrl *msm_pinctrl_data;
+atomic_t skip_eoi[NR_CPUS];
 
 #define MSM_ACCESSOR(name) \
 static u32 msm_readl_##name(struct msm_pinctrl *pctrl, \
@@ -1003,6 +1004,17 @@ static void msm_gpio_irq_ack(struct irq_data *d)
 	raw_spin_unlock_irqrestore(&pctrl->lock, flags);
 }
 
+static void msm_gpio_irq_eoi(struct irq_data *d)
+{
+	int cpu = raw_smp_processor_id();
+
+	if (atomic_read(&skip_eoi[cpu]) == 1) {
+		atomic_set(&skip_eoi[cpu], 0);
+		return;
+	}
+	irq_chip_eoi_parent(d);
+}
+
 static void msm_dirconn_cfg_reg(struct irq_data *d, u32 offset)
 {
 	u32 val;
@@ -1077,11 +1089,13 @@ static void msm_gpio_dirconn_handler(struct irq_desc *desc)
 {
 	struct irq_data *irqd = irq_desc_get_handler_data(desc);
 	struct irq_chip *chip = irq_desc_get_chip(desc);
+	int cpu = raw_smp_processor_id();
 
 	if (!irqd)
 		return;
 
 	chained_irq_enter(chip, desc);
+	atomic_set(&skip_eoi[cpu], 1);
 	generic_handle_irq(irqd->irq);
 	chained_irq_exit(chip, desc);
 	irq_set_irqchip_state(irq_desc_get_irq_data(desc)->irq,
@@ -1465,7 +1479,7 @@ static int msm_gpio_init(struct msm_pinctrl *pctrl)
 	pctrl->irq_chip.irq_mask = msm_gpio_irq_mask;
 	pctrl->irq_chip.irq_unmask = msm_gpio_irq_unmask;
 	pctrl->irq_chip.irq_ack = msm_gpio_irq_ack;
-	pctrl->irq_chip.irq_eoi = irq_chip_eoi_parent;
+	pctrl->irq_chip.irq_eoi = msm_gpio_irq_eoi;
 	pctrl->irq_chip.irq_set_type = msm_gpio_irq_set_type;
 	pctrl->irq_chip.irq_set_wake = msm_gpio_irq_set_wake;
 	pctrl->irq_chip.irq_set_affinity = msm_gpio_irq_set_affinity;
@@ -1877,6 +1891,9 @@ int msm_pinctrl_probe(struct platform_device *pdev,
 
 	raw_spin_lock_init(&pctrl->lock);
 
+	for (i = 0; i < NR_CPUS; i++)
+		atomic_set(&skip_eoi[i], 0);
+
 	if (soc_data->tiles) {
 		for (i = 0; i < soc_data->ntiles; i++) {
 			res = platform_get_resource_byname(pdev, IORESOURCE_MEM,
-- 
2.49.0

